File: sdk/go/dagger/querybuilder/querybuilder_test.go
package querybuilder

import (
	"encoding/json"
	"testing"

	"github.com/stretchr/testify/require"
)

func TestQuery(t *testing.T) {
	root := Query().
		Select("core").
		Select("image").Arg("ref", "alpine").
		Select("file").Arg("path", "/etc/alpine-release")

	q := root.Build()
	require.Equal(t, `query{core{image(ref:"alpine"){file(path:"/etc/alpine-release")}}}`, q)
}

func TestAlias(t *testing.T) {
	root := Query().
		Select("core").
		Select("image").Arg("ref", "alpine").
		SelectWithAlias("foo", "file").Arg("path", "/etc/alpine-release")

	q := root.Build()
	require.Equal(t, `query{core{image(ref:"alpine"){foo:file(path:"/etc/alpine-release")}}}`, q)
}

func TestArgsCollision(t *testing.T) {
	q := Query().
		Select("a").Arg("arg", "one").
		Select("b").Arg("arg", "two").
		Build()
	require.Equal(t, `query{a(arg:"one"){b(arg:"two")}}`, q)
}

func TestNullableArgs(t *testing.T) {
	str := "value"

	tests := []struct {
		arg    any
		expect string
	}{
		{
			expect: `query{a(arg:"value")}`,
			arg:    str,
		},
		{
			expect: `query{a(arg:"value")}`,
			arg:    &str,
		},
		{
			expect: `query{a(arg:["value"])}`,
			arg:    []string{str},
		},
		{
			expect: `query{a(arg:["value"])}`,
			arg:    []*string{&str},
		},
		{
			expect: `query{a(arg:["value"])}`,
			arg:    &([]*string{&str}),
		},
	}

	for _, test := range tests {
		q := Query().Select("a").Arg("arg", test.arg).Build()
		require.Equal(t, test.expect, q)
	}
}

func TestFieldImmutability(t *testing.T) {
	root := Query().
		Select("test")

	a := root.Select("a").Build()
	require.Equal(t, `query{test{a}}`, a)

	// Make sure this is not `test{a,b}` (e.g. the previous select didn't modify `root` in-place)
	b := root.Select("b").Build()
	require.Equal(t, `query{test{b}}`, b)
}

func TestArgImmutability(t *testing.T) {
	root := Query().
		Select("test")

	a := root.Arg("foo", "bar").Build()
	require.Equal(t, `query{test(foo:"bar")}`, a)

	// Make sure this does not contain `hello` (e.g. the previous select didn't modify `root` in-place)
	b := root.Arg("hello", "world").Build()
	require.Equal(t, `query{test(hello:"world")}`, b)
}

func TestUnpack(t *testing.T) {
	var contents string
	root := Query().
		Select("foo").
		Select("bar").Arg("hello", "world").
		Select("field").Arg("test", "test").Bind(&contents)

	var response any
	err := json.Unmarshal([]byte(`
		{
			"foo": {
				"bar": {
					"field": "TEST"
				}
			}
		}
	`), &response)
	require.NoError(t, err)
	require.NoError(t, root.Unpack(response))
	require.Equal(t, "TEST", contents)
}

File: sdk/elixir/test/dagger/querybuilder_test.exs
defmodule Dagger.QueryBuilder.SelectionTest do
  use ExUnit.Case, async: true

  alias Dagger.QueryBuilder.Selection

  test "query" do
    root =
      Selection.query()
      |> Selection.select("core")
      |> Selection.select("image")
      |> Selection.arg("ref", "alpine")
      |> Selection.select("file")
      |> Selection.arg("path", "/etc/alpine-release")

    assert Selection.build(root) ==
             "query{core{image(ref:\"alpine\"){file(path:\"/etc/alpine-release\")}}}"
  end

  test "alias" do
    root =
      Selection.query()
      |> Selection.select("core")
      |> Selection.select("image")
      |> Selection.arg("ref", "alpine")
      |> Selection.select_with_alias("foo", "file")
      |> Selection.arg("path", "/etc/alpine-release")

    assert Selection.build(root) ==
             "query{core{image(ref:\"alpine\"){foo:file(path:\"/etc/alpine-release\")}}}"
  end

  test "select multi fields" do
    root =
      Selection.query()
      |> Selection.select("core")
      |> Selection.select("name value")

    assert Selection.build(root) == "query{core{name value}}"
  end

  test "args" do
    root =
      Selection.query()
      |> Selection.select("a")
      |> Selection.arg("arg", "b")
      |> Selection.arg("arg1", "c")

    assert Selection.build(root) == "query{a(arg:\"b\",arg1:\"c\")}"
  end

  test "arg collision" do
    root =
      Selection.query()
      |> Selection.select("a")
      |> Selection.arg("arg", "one")
      |> Selection.select("b")
      |> Selection.arg("arg", "two")

    assert Selection.build(root) == "query{a(arg:\"one\"){b(arg:\"two\")}}"
  end

  test "array args" do
    root =
      Selection.query()
      |> Selection.select("a")
      |> Selection.arg("arg", [])

    assert Selection.build(root) == "query{a(arg:[])}"

    root =
      Selection.query()
      |> Selection.select("a")
      |> Selection.arg("arg", ["value"])

    assert Selection.build(root) == "query{a(arg:[\"value\"])}"

    root =
      Selection.query()
      |> Selection.select("a")
      |> Selection.arg("arg", ["value", "value2"])

    assert Selection.build(root) == "query{a(arg:[\"value\",\"value2\"])}"

    root =
      Selection.query()
      |> Selection.select("a")
      |> Selection.arg("arg", [%{"name" => "foo"}])

    assert Selection.build(root) == "query{a(arg:[{name:\"foo\"}])}"
  end

  test "object args" do
    root =
      Selection.query()
      |> Selection.select("a")
      |> Selection.arg("arg", %{"name" => "a", "value" => "b"})

    assert Selection.build(root) == "query{a(arg:{name:\"a\",value:\"b\"})}"
  end

  test "string arg escape" do
    root =
      Selection.query()
      |> Selection.select("a")
      |> Selection.arg("arg", "\n\t\"")

    assert Selection.build(root) == "query{a(arg:\"\\n\\t\\\"\")}"
  end

  test "boolean arg" do
    root =
      Selection.query()
      |> Selection.select("a")
      |> Selection.arg("arg", true)

    assert Selection.build(root) == "query{a(arg:true)}"
  end
end
